Uniquify all Sprite preterms in separate step prior elaboration

janvrany opened the pull request:
Yet another attempt to fix #139.

This PR α-renames all variables in a separate step prior elaboration.
It is similar to PR #197 but differ in several ways:

 * perform renaming as a separate step, not as part of [Chk-Lam].
 * renames all lets, not just function parameter binds
 * renames everything in single top-down traversal.

janvrany committed:
Add one more test for Issue #139

janvrany committed:
Uniquify all Sprite preterms before elaboartion

This commit α-renames all variables in a separate step prior elaboration
in order to avoid name clashes as demonstrated by `#testAlphaClash`.

This is similar to PR #197 but differ in several ways:

 * perform renaming as a separate step, not as part of [Chk-Lam].
 * renames all lets, not just function parameter binds
 * renames everything in single top-down traversal.

Fixes #139 (well, not quite yet)

shingarov commented:
I wonder if this can be somehow integrated with `AlphaRenamer`.  Which leads us to further zoom in attention to things like #194 , which is just one manifestation of a more general issue, that in the upstream Sprite, there is a hierarchy of Exprs parallel to that in Fixpoint.  For example, there are `EVar` and `EApp` etc in Fixpoint, and there are `EVar` and `EApp` etc in Sprite (and in the Sprite namespace the former are called `F.EVar` etc because of `import qualified Language.Fixpoint.Types as F`).  Actually it's even worse: in Fixpoint itself there is duplication between the "fixpoint names" (`F.*`) and "horn names" (`H.*`).  In addition to datatype names, there is similar duplication of function names.

In Smalltalk, most of the time we follow the convention that if some `a` is duplicated in Sprite and Fixpoint, the Fixpoint thing (`F.a`) is called `a` in Smalltalk, and the Sprite `a` is called `Λa` in Smalltalk.  See, for example, `Object>>predReft`, or `SInfo>>eliminatingSolverInfo`.

But sometimes it's too unbearable to lug all this parallelism through the whole codebase.  There is, for example, `PInt` ("literal integer constant").  I just don't bother distinguishing between the Sprite and the Fixpoint `PInt`.  One day a hero will be born who will clean all this mess.

janvrany commented:
> I wonder if this can be somehow integrated with AlphaRenamer.

I was thinking the same but after digging deeper I decided not to, not now anyways. It seemed more than I could chew. 

> One day a hero will be born who will clean all this mess.

Yeah, and hopefully soon else a mere hero won't suffice. And gods are even scarcer to come by than heroes.

janvrany committed:
Fix bug in type renaming w.r.t termination metric

Both MA (prior this commit) and upstream Sprite (at the time of this
commit) had bug in type renaming where parameters were renamed but
not references to them in termination metric. All tests "appeared to
work" simply because the code used same (formal) parameter names in both
type signature and function declaration.

However in general case they do not need to be the same. Solution is
simple - also rename parameters in termination metric.

janvrany commented:
Closing as resolved via 3ac081f and #283
